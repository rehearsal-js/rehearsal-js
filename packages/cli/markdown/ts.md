`rehearsal ts`
==============

bump typescript dev-dependency with compilation insights and auto-fix

* [`rehearsal ts`](#rehearsal-ts)

## `rehearsal ts`

bump typescript dev-dependency with compilation insights and auto-fix

```
USAGE
  $ rehearsal ts

OPTIONS
  -b, --build=beta|next|latest  [default: beta] typescript build variant
  -s, --src_dir=src_dir         [default: ./app] typescript source directory

ALIASES
  $ rehearsal typescript
```

_See code: [dist/src/commands/ts.ts](https://github.com/rehearsal-js/rehearsal-js/blob/v0.0.1/dist/src/commands/ts.ts)_
