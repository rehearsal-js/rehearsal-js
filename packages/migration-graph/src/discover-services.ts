import { parseSync } from '@swc/core';
import debug from 'debug';

import type {
  CallExpression,
  ClassDeclaration,
  ClassExpression,
  ClassMember,
  ClassProperty,
  Decorator,
  ExportDefaultDeclaration,
  Identifier,
  ImportDeclaration,
  ImportSpecifier,
  ModuleItem,
  NamedImportSpecifier,
  StringLiteral,
} from '@swc/core';
import type { CustomImportResolver } from './types.js';

const DEBUG_CALLBACK = debug('rehearsal:migration-graph:discover-ember-service-dependencies');

export function discoverServiceDependencies(
  serviceMap: Record<string, string>
): CustomImportResolver {
  return (contentType: 'ecmascript' | 'typescript', content: string): string[] => {
    if (!(content.includes('@ember/service') || content.includes('@glimmerx/service'))) {
      return [];
    }

    const parsed = parseSync(content, {
      syntax: contentType,
      decorators: true,
      decoratorsBeforeExport: true,
    });

    if (!parsed) {
      return [];
    }

    // Get all imports of '@ember/service'.

    const maybeImportDeclarations = filterImportDeclarationsBySource(parsed.body, '@ember/service');

    if (!maybeImportDeclarations) {
      DEBUG_CALLBACK('No import declaration found');
      return [];
    }

    // We need to look through all import declarations to find for exports with either:
    // `inject` 3.28
    // `service` 4+

    const foundImportDeclarationUsingInject = findImportDeclarationWithExportedName(
      maybeImportDeclarations,
      'inject'
    );

    const foundImportDeclarationUsingService = findImportDeclarationWithExportedName(
      maybeImportDeclarations,
      'service'
    );

    const importDeclaration =
      foundImportDeclarationUsingInject || foundImportDeclarationUsingService;

    if (!importDeclaration) {
      DEBUG_CALLBACK('No import declaration found with exported names: inject or service');
      return [];
    }

    // In the case of re-assignment from inject as service, we walk the specifiers
    const maybeInjectSpecifier = findSpecifierByExportedName(
      importDeclaration.specifiers,
      'inject'
    );
    const maybeServiceSpecifier = findSpecifierByExportedName(
      importDeclaration.specifiers,
      'service'
    );

    const maybeSpecifier = maybeInjectSpecifier || maybeServiceSpecifier;

    if (!maybeSpecifier) {
      return [];
    }

    const decoratorName = maybeSpecifier.local.value;

    if (!decoratorName) {
      return [];
    }

    let foundClasses: Array<ClassExpression | ClassDeclaration> = [];

    // Find all ClassDeclaration in body of the parsed file
    const classDeclarations = findClassDeclarations(parsed.body);

    if (classDeclarations) {
      foundClasses = foundClasses.concat(Array.from(classDeclarations));
    }

    // Typical use case will have ExportDefaultDeclaration where the body is a ClassExpression
    const someDefaultExport = parsed.body.find(
      (m: ModuleItem) => m.type == 'ExportDefaultDeclaration'
    );

    if (someDefaultExport) {
      const edd = someDefaultExport as ExportDefaultDeclaration;
      // Check if the ExportDefaultDeclaration has a ClassExpression
      if (edd.decl.type === 'ClassExpression') {
        foundClasses.push(edd.decl);
      } else {
        DEBUG_CALLBACK('Default export does not contain a ClassExpression');
      }
    }

    if (foundClasses.length < 1) {
      return [];
    }

    let results: string[] = [];

    for (const classExp of foundClasses) {
      const body = classExp.body;

      const classProperties = filterByClassPropertyWithDecorators(body);

      if (!classProperties) {
        continue;
      }

      const classPropertiesWithDecorators = filterClassPropertyWithDecoratorName(
        classProperties,
        decoratorName
      );

      if (!classPropertiesWithDecorators) {
        continue;
      }

      results = classPropertiesWithDecorators.map((prop: ClassProperty) => {
        if (prop.decorators) {
          const maybeDecorator = findDecoratorWithName(prop.decorators, decoratorName);

          if (!!maybeDecorator && isDecoratorWithExpressionTypeCallExpression(maybeDecorator)) {
            const callExpression = maybeDecorator.expression as CallExpression;
            const arg = callExpression.arguments[0].expression as StringLiteral;
            // someAddon@serviceName

            if (serviceMap[arg.value]) {
              return serviceMap[arg.value];
            }

            return parseServiceMetaFromString(arg.value);
          }
        }

        const keyName = (prop.key as Identifier).value;

        if (serviceMap[keyName]) {
          return serviceMap[keyName];
        }

        // Fallback if decorator.expression where decoratorName`  is not call expression
        return intoFileName(keyName);
      });
    }

    return results;
  };
}

function findClassDeclarations(items: ModuleItem[]): Array<ClassDeclaration> | undefined {
  return items
    .filter((i: ModuleItem) => i.type == 'ClassDeclaration')
    .map((i) => i as ClassDeclaration);
}

function parseServiceMetaFromString(str: string): string {
  if (str.includes('@')) {
    const idx = str.lastIndexOf('@');
    const packageNameOrModuleName = str.substring(0, idx);
    const serviceName = str.substring(idx + 1);
    return `${packageNameOrModuleName}/services/${intoFileName(serviceName)}`;
  }
  return intoFileName(str);
}

const STRING_DECAMELIZE_REGEXP = /([a-z\d])([A-Z])/g;
const STRING_DASHERIZE_REGEXP = /[ _]/g;
function intoFileName(str: string): string {
  return str
    .replace(STRING_DECAMELIZE_REGEXP, '$1_$2')
    .toLowerCase()
    .replace(STRING_DASHERIZE_REGEXP, '-');
}

function filterByClassPropertyWithDecorators(members: ClassMember[]): ClassProperty[] {
  const result: unknown = members.filter(
    (m: ClassMember) => m.type == 'ClassProperty' && !!m.decorators
  );
  return result as ClassProperty[];
}

function filterClassPropertyWithDecoratorName(
  classProperties: ClassProperty[],
  decoratorName: string
): ClassProperty[] {
  return classProperties.filter((prop: ClassProperty) => {
    const decorators = prop.decorators;
    return !!decorators && hasDecoratorWithName(decorators, decoratorName);
  });
}

function findDecoratorWithName(
  decorators: Decorator[],
  decoratorName: string
): Decorator | undefined {
  return decorators.find((d) => {
    let identifier: Identifier | undefined;

    // For a given decorate its expression property can either
    // be an Identifier or CallExpression

    // If it's an identifier it would look like
    // e.g. `@service name;`
    if (isDecoratorWithExpressionTypeIdentifier(d)) {
      identifier = d.expression as Identifier;
    }

    // If it's a CallExpression this is the case where we pass metadata
    // into the decorator
    // e.g. `@service('someAddon@serviceName') myVariable;`
    else if (isDecoratorWithExpressionTypeCallExpression(d)) {
      const callExpression = d.expression as CallExpression;
      identifier = callExpression.callee as Identifier;
    }

    return !!identifier && identifier.value === decoratorName;
  });
}

function hasDecoratorWithName(decorators: Decorator[], decoratorName: string): boolean {
  return !!findDecoratorWithName(decorators, decoratorName);
}

function isDecoratorWithExpressionTypeIdentifier(d: Decorator): boolean {
  return d.expression.type === 'Identifier';
}

function isDecoratorWithExpressionTypeCallExpression(d: Decorator): boolean {
  return d.expression.type === 'CallExpression';
}

function filterImportDeclarationsBySource(
  items: ModuleItem[],
  source: string
): ImportDeclaration[] {
  return items.filter((i) => isImportDeclaration(i, source)).map((i) => i as ImportDeclaration);
}

function findImportDeclarationWithExportedName(
  items: ImportDeclaration[],
  exportName: string
): ImportDeclaration | undefined {
  return items.find((i: ImportDeclaration) => {
    const found = findSpecifierByExportedName(i.specifiers, exportName);

    return found ?? false;
  });
}

function isImportDeclaration(m: ModuleItem, source: string): boolean {
  return m.type === 'ImportDeclaration' && m.source.value === source;
}

function findSpecifierByExportedName(
  specifiers: ImportSpecifier[],
  exportName: string
): ImportSpecifier | undefined {
  return specifiers.find((s: NamedImportSpecifier | ImportSpecifier) => {
    // If imported is null, then only have a local value, which implies no re-assignment.
    // e.g.
    // import { foo as bar } from 'baz'; -> bar is local, foo is imported
    // import { foo } from 'baz'; -> foo is local, imported is null
    return (
      (s as NamedImportSpecifier)?.imported?.value === exportName || s.local.value === exportName
    );
  });
}
