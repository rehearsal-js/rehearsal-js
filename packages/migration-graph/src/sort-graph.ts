import type { FileNode } from './file-node.js';
import type { PackageGraph } from './project-graph.js';

export function topSortFiles(graph: PackageGraph): FileNode[] {
  const nodes: FileNode[] = [];

  // Mark all nodes as unvisited
  const visited: Map<string, boolean> = new Map();
  for (const pkg of graph.packages.values()) {
    for (const file of pkg.files) {
      visited.set(file.id, false);
      nodes.push(file);
    }
  }

  const result: FileNode[] = [];

  // Helper function to visit a node and its children recursively
  function visit(node: FileNode): void {
    if (!visited.get(node.id)) {
      visited.set(node.id, true);
      for (const child of node.edges) {
        visit(child);
      }
      result.push(node);
    }
  }

  // Visit each unvisited node
  for (const node of nodes) {
    visit(node);
  }

  return result;
}
