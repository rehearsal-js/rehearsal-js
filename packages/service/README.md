# @rehearsal/service

Rehearsal Service provides access to project's files and helps to run plugins to transform their source code.
