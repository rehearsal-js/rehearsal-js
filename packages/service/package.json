{
  "name": "@rehearsal/service",
  "version": "1.0.9-beta",
  "description": "Rehearsal Service",
  "keywords": [
    "rehearsal",
    "service",
    "ast",
    "transforms",
    "typescript",
    "glint"
  ],
  "repository": {
    "type": "git",
    "url": "https://github.com/rehearsal-js/rehearsal-js.git"
  },
  "license": "BSD-2-Clause",
  "main": "dist/src/index.js",
  "types": "dist/src/index.d.ts",
  "files": [
    "dist"
  ],
  "scripts": {
    "build": "tsc -b",
    "lint": "npm-run-all lint:*",
    "lint:eslint": "eslint --fix . --ext .ts",
    "lint:tsc-src": "tsc --noEmit",
    "lint:tsc-test": "tsc --noEmit --project test/tsconfig.json",
    "test": "vitest --run",
    "test:watch": "vitest --coverage --watch",
    "version": "pnpm version"
  },
  "dependencies": {
    "@glint/core": "^1.0.0-beta.3",
    "@rehearsal/reporter": "workspace:*",
    "@rehearsal/utils": "workspace:*",
    "findup-sync": "^5.0.0",
    "winston": "^3.8.2"
  },
  "devDependencies": {
    "vitest": "^0.28.4"
  },
  "peerDependencies": {
    "typescript": "^4.9 || ^5.0"
  },
  "packageManager": "pnpm@7.12.1",
  "engines": {
    "node": ">=14.16.0"
  },
  "volta": {
    "extends": "../../package.json"
  }
}
