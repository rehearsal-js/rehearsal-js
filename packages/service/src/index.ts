export { RehearsalService as default } from './rehearsal-service';

export { Plugin } from './plugin';
export { RehearsalService } from './rehearsal-service';
export { RehearsalServiceHost } from './rehearsal-service-host';

export type { PluginOptions, PluginParams, PluginResult } from './plugin';
