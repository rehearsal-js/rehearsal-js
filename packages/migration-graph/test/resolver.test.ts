import { describe, beforeEach, expect, test, afterEach } from 'vitest';
import { Project } from 'fixturify-project';
import { Resolver } from '../src/resolver.js';
import { discoverServiceDependencies } from '../src/discover-services.js';
import { topSortFiles } from '../src/sort-graph.js';

describe('Resolver', () => {
  let resolver: Resolver;
  let project: Project;

  beforeEach(async () => {
    resolver = new Resolver();
    await resolver.load();
    project = new Project('package-a');
    project.files = {
      'tsconfig.json': JSON.stringify({
        compilerOptions: {
          baseUrl: './',
        },
        references: [
          { path: './packages/package-a' },
          { path: './packages/package-b' },
          { path: './packages/package-c' },
        ],
      }),
      packages: {
        'package-a': {
          src: {
            'a.js': `
            import { b } from "./b.js";
            import { index } from './thing';
            export const a = "a";`,
            'b.js': `
            import { c } from './c.js';
            export const b = "b";
            `,
            'c.js': `
            import { a } from "./a.js";
            export const c = "c";`,
            thing: {
              'index.ts': `export const index = "index";`,
            },
          },
          'package.json': JSON.stringify({
            name: 'package-a',
          }),
          'tsconfig.json': JSON.stringify({
            extends: '../../tsconfig.json',
          }),
        },
        'package-b': {
          src: {
            'd.js': `
            import { b } from "package-a/b";
            import { e } from "./e.ts";
            const d = "d";`,
            'e.ts': `
            export const e = "e";
            `,
            'f.js': 'import { e } from "./e.js";\nexport const f = "f";',
          },
          'package.json': JSON.stringify({
            name: 'package-b',
            dependencies: {
              'package-a': '*',
            },
          }),
          'tsconfig.json': JSON.stringify({
            extends: '../../tsconfig.json',
            compilerOptions: {
              paths: {
                'package-a/*': ['./packages/package-a/src/*'],
              },
            },
          }),
        },
        'package-c': {
          src: {
            'g.gts': `
            import Component from "@glimmer/component";
            import { h } from './h.js';
            import { f } from 'package-b/f';

            export default class G extends Component {
              <template>{{@name}}</template>
            }
            `,
            'h.ts': `export const h = "h";`,
            'j.gjs': `
            import { h } from './h.js';
            import { f } from 'package-b/f';
            import Component from "@glimmer/component";
            import { inject as service } from "@ember/service";
            import K from './k.gjs';

            export default class J extends Component {
              @service('package-d@thing') thing;
              @service tracking;
              @service('nothing@not-found') nothing;
              <template>
                <K />
              </template>
            }
            `,
            'k.gjs': `
            import Component from "@glimmer/component";

            export default class K extends Component {
              <template>
                K
              </template>
            }
            `,
          },
          'package.json': JSON.stringify({
            name: 'package-c',
            dependencies: {
              'package-b': '*',
              'package-d': '*',
            },
          }),
          'tsconfig.json': JSON.stringify({
            extends: '../../tsconfig.json',
            compilerOptions: {
              paths: {
                'package-b/*': ['./packages/package-b/src/*'],
                'package-d/*': ['./packages/package-d/addon/*'],
              },
            },
          }),
        },
        'package-d': {
          addon: {
            services: {
              'thing.js': `
              import Service from "@ember/service";

              export class Thing extends Service {}
              `,

              'tracking.js': `
              import Service from "@ember/service";

              export class Tracking extends Service {}
              `,
            },
          },
          'package.json': JSON.stringify({
            name: '@company/package-d',
            dependencies: {
              'package-b': '*',
            },
          }),
          'tsconfig.json': JSON.stringify({
            extends: '../../tsconfig.json',
            compilerOptions: {
              paths: {
                'package-b/*': ['./packages/package-b/src/*'],
              },
            },
          }),
        },

        'package-e': {
          'dist-types': {
            src: {
              foo: {
                'index.d.ts': 'export declare const index: string;',
              },
            },
          },
          src: {
            '1.tsx': `
            import { two } from './2.mts';
            import { Three } from './3';
            import { index } from './foo';
            two;
            Three;
            index;
            `,
            '2.mts': `export const two = "two";`,
            '3.jsx': `
            import { b } from 'package-a/b';
            b;
            import { Four } from './4.jsx';
            export function Three() { return <div>Hello <Four /></div> }
            `,
            '4.jsx': `export function Four() { return <div>Hello</div> }`,
            foo: {
              'index.ts': `export const index = "index";`,
            },
          },
          'package.json': JSON.stringify({
            name: '@company/package-e',
            dependencies: {
              'package-b': '*',
            },
          }),
          'tsconfig.json': JSON.stringify({
            extends: '../../tsconfig.json',
            compilerOptions: {
              paths: {
                'package-e/*': ['./packages/package-e/src/*'],
                'package-a/*': ['./packages/package-a/src/*'],
              },
            },
          }),
        },

        'package-f': {
          src: {
            'index.mts': `import {index} from "package-e/foo"`,
          },
          'package.json': JSON.stringify({
            name: '@company/package-4',
            dependencies: {
              'package-e': '*',
            },
          }),
          'tsconfig.json': JSON.stringify({
            extends: '../../tsconfig.json',
            compilerOptions: {
              paths: {
                'package-e/*': ['./packages/package-e/dist-types/src/*'],
              },
            },
          }),
        },
      },
    };

    await project.write();
  });

  afterEach(() => {
    project.dispose();
  });

  test('it works', () => {
    resolver.walk(project.baseDir + '/packages/package-b/src/d.js');

    expect(topSortFiles(resolver.graph)).toMatchObject([
      project.baseDir + '/packages/package-b/src/e.ts',
      project.baseDir + '/packages/package-a/src/thing/index.ts',
      project.baseDir + '/packages/package-a/src/a.js',
      project.baseDir + '/packages/package-a/src/c.js',
      project.baseDir + '/packages/package-a/src/b.js',
      project.baseDir + '/packages/package-b/src/d.js',
    ]);
  });

  test('it works gts', () => {
    resolver.walk(project.baseDir + '/packages/package-c/src/g.gts');

    expect(topSortFiles(resolver.graph)).toMatchObject([
      project.baseDir + '/packages/package-b/src/e.ts',
      project.baseDir + '/packages/package-b/src/f.js',
      project.baseDir + '/packages/package-c/src/h.ts',
      project.baseDir + '/packages/package-c/src/g.gts',
    ]);
  });

  test('it works tsx', () => {
    resolver.walk(project.baseDir + '/packages/package-e/src/1.tsx');

    expect(topSortFiles(resolver.graph)).toMatchObject([
      project.baseDir + '/packages/package-e/src/foo/index.ts',
      project.baseDir + '/packages/package-e/src/4.jsx',
      project.baseDir + '/packages/package-a/src/thing/index.ts',
      project.baseDir + '/packages/package-a/src/a.js',
      project.baseDir + '/packages/package-a/src/c.js',
      project.baseDir + '/packages/package-a/src/b.js',
      project.baseDir + '/packages/package-e/src/3.jsx',
      project.baseDir + '/packages/package-e/src/2.mts',
      project.baseDir + '/packages/package-e/src/1.tsx',
    ]);
  });

  test('can find service when given a resolver', () => {
    const resolveServices = discoverServiceDependencies({});

    resolver = new Resolver({ customResolver: resolveServices });

    resolver.walk(project.baseDir + '/packages/package-c/src/j.gjs');
    expect(topSortFiles(resolver.graph)).toMatchObject([
      project.baseDir + '/packages/package-d/addon/services/thing.js',
      project.baseDir + '/packages/package-c/src/k.gjs',
      project.baseDir + '/packages/package-b/src/e.ts',
      project.baseDir + '/packages/package-b/src/f.js',
      project.baseDir + '/packages/package-c/src/h.ts',
      project.baseDir + '/packages/package-c/src/j.gjs',
    ]);
  });

  test('can find services when given a resolver with a map', () => {
    const resolveServices = discoverServiceDependencies({
      tracking: 'package-d/services/tracking.js',
    });

    resolver = new Resolver({ customResolver: resolveServices });

    resolver.walk(project.baseDir + '/packages/package-c/src/j.gjs');

    expect(topSortFiles(resolver.graph)).toMatchObject([
      project.baseDir + '/packages/package-d/addon/services/tracking.js',
      project.baseDir + '/packages/package-d/addon/services/thing.js',
      project.baseDir + '/packages/package-c/src/k.gjs',
      project.baseDir + '/packages/package-b/src/e.ts',
      project.baseDir + '/packages/package-b/src/f.js',
      project.baseDir + '/packages/package-c/src/h.ts',
      project.baseDir + '/packages/package-c/src/j.gjs',
    ]);
  });

  test('can resolve up from a declaration file if it has paths to itself', () => {
    resolver.walk(project.baseDir + '/packages/package-f/src/index.mts');

    expect(topSortFiles(resolver.graph)).toMatchObject([
      project.baseDir + '/packages/package-e/src/foo/index.ts',
      project.baseDir + '/packages/package-f/src/index.mts',
    ]);
  });
});
