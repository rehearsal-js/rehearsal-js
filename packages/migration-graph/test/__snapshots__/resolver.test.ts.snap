// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Resolver > resolver produces a graph that can be emitted as dotlang 1`] = `
"strict digraph {

  rankdir=\\"LR\\" splines=\\"true\\" overlap=\\"false\\" nodesep=\\"0.16\\" ranksep=\\"0.18\\" fontname=\\"Helvetica-bold\\" fontsize=\\"9\\" style=\\"rounded,bold,filled\\" fillcolor=\\"#ffffff\\" compound=\\"true\\"
  node [shape=\\"box\\" style=\\"rounded, filled\\" height=\\"0.2\\" color=\\"black\\" fillcolor=\\"#ffffcc\\" fontcolor=\\"black\\" fontname=\\"Helvetica\\" fontsize=\\"9\\"]
  edge [arrowhead=\\"normal\\" arrowsize=\\"0.6\\" penwidth=\\"2.0\\" color=\\"#00000033\\" fontname=\\"Helvetica\\" fontsize=\\"9\\"]
    subgraph cluster_companypackage_with_externals {
    label=\\"@company/package-with-externals\\";
    href=\\"file://<tmp-path>/packages/with-externals-imports/package.json\\";
    labeljust=l;
    color=black;
    node [style=filled, fillcolor=white, shape=box];
    \\"<tmp-path>/packages/with-externals-imports/src/index.mts\\" [label=\\"src/index.mts\\", href=\\"file://<tmp-path>/packages/with-externals-imports/src/index.mts\\"];
    \\"<tmp-path>/packages/with-externals-imports/src/other.js\\" [label=\\"src/other.js\\", href=\\"file://<tmp-path>/packages/with-externals-imports/src/other.js\\"];
  }
  subgraph cluster_companypackage_e {
    label=\\"@company/package-e\\";
    href=\\"file://<tmp-path>/packages/package-e/package.json\\";
    labeljust=l;
    color=black;
    node [style=filled, fillcolor=white, shape=box];
    \\"<tmp-path>/packages/package-e/src/foo/index.ts\\" [label=\\"src/foo/index.ts\\", href=\\"file://<tmp-path>/packages/package-e/src/foo/index.ts\\"];
  }
  subgraph cluster_my_external {
    label=\\"my-external\\";
    labeljust=l;
    href=\\"file://<tmp-path>/packages/with-externals-imports/node_modules/my-external/package.json\\";
    color=\\"grey\\"
    style=\\"dashed\\"
    fontcolor=\\"grey\\"
    node [style=filled, fillcolor=white, shape=box];
    \\"<tmp-path>/packages/with-externals-imports/node_modules/my-external/dist/index.js\\" [label=\\"dist/index.js\\"];
  }
  subgraph cluster_my_external_addon_no_types {
    label=\\"my-external-addon-no-types\\";
    labeljust=l;
    href=\\"file://<tmp-path>/packages/with-externals-imports/node_modules/my-external-addon-no-types/package.json\\";
    color=\\"grey\\"
    style=\\"dashed\\"
    fontcolor=\\"grey\\"
    node [style=filled, fillcolor=white, shape=box];
    \\"<tmp-path>/packages/with-externals-imports/node_modules/my-external-addon-no-types/addon/index.js\\" [label=\\"addon/index.js\\"];
  }
  subgraph cluster_my_external_addon_no_types {
    label=\\"my-external-addon-no-types\\";
    labeljust=l;
    href=\\"file://<tmp-path>/packages/with-externals-imports/node_modules/my-external-addon-no-types/package.json\\";
    color=\\"grey\\"
    style=\\"dashed\\"
    fontcolor=\\"grey\\"
    node [style=filled, fillcolor=white, shape=box];
    \\"<tmp-path>/packages/with-externals-imports/node_modules/my-external-addon-no-types/addon/index.js\\" [label=\\"addon/index.js\\"];
  }
  subgraph cluster_my_external_with_types {
    label=\\"my-external-with-types\\";
    labeljust=l;
    href=\\"file://<tmp-path>/packages/with-externals-imports/node_modules/my-external-with-types/package.json\\";
    color=\\"grey\\"
    style=\\"dashed\\"
    fontcolor=\\"grey\\"
    node [style=filled, fillcolor=white, shape=box];
    \\"<tmp-path>/packages/with-externals-imports/node_modules/my-external-with-types/addon/index.ts\\" [label=\\"addon/index.ts\\"];
  }
  subgraph cluster_my_external_with_types {
    label=\\"my-external-with-types\\";
    labeljust=l;
    href=\\"file://<tmp-path>/packages/with-externals-imports/node_modules/my-external-with-types/package.json\\";
    color=\\"grey\\"
    style=\\"dashed\\"
    fontcolor=\\"grey\\"
    node [style=filled, fillcolor=white, shape=box];
    \\"<tmp-path>/packages/with-externals-imports/node_modules/my-external-with-types/addon/index.ts\\" [label=\\"addon/index.ts\\"];
  }
  subgraph cluster_emberobject {
    label=\\"@ember/object\\";
    labeljust=l;
    href=\\"file://node_modules/missing/package.json\\";
    color=\\"grey\\"
    style=\\"dashed\\"
    fontcolor=\\"grey\\"
    node [style=filled, fillcolor=white, shape=box];
    \\"@ember/object\\" [label=\\"../../@ember/object\\"];
  }
  \\"<tmp-path>/packages/with-externals-imports/src/index.mts\\" -> \\"<tmp-path>/packages/with-externals-imports/src/other.js\\" [xlabel=\\"missing types\\" tooltip=\\"missing types\\" fontcolor=\\"red\\" color=\\"red\\"];
  \\"<tmp-path>/packages/with-externals-imports/src/index.mts\\" -> \\"<tmp-path>/packages/with-externals-imports/node_modules/my-external-addon-no-types/addon/index.js\\" [xlabel=\\"missing types\\" tooltip=\\"missing types\\" fontcolor=\\"red\\" color=\\"red\\"];
  \\"<tmp-path>/packages/with-externals-imports/src/index.mts\\" -> \\"<tmp-path>/packages/with-externals-imports/node_modules/my-external/dist/index.js\\" [color=\\"grey\\" style=\\"dashed\\"];
  \\"<tmp-path>/packages/with-externals-imports/src/index.mts\\" -> \\"<tmp-path>/packages/package-e/src/foo/index.ts\\" [xlabel=\\"missing from package.json\\" tooltip=\\"missing from package.json\\" fontcolor=\\"orange\\" color=\\"orange\\"];
  \\"<tmp-path>/packages/with-externals-imports/src/other.js\\" -> \\"@ember/object\\" [xlabel=\\"missing from package.json\\" tooltip=\\"missing from package.json\\" fontcolor=\\"orange\\" color=\\"orange\\"];
  \\"<tmp-path>/packages/with-externals-imports/src/other.js\\" -> \\"<tmp-path>/packages/with-externals-imports/node_modules/my-external-with-types/addon/index.ts\\" [color=\\"grey\\" style=\\"dashed\\"];
}"
`;

exports[`Resolver > resolver produces a graph that can be emitted as json 1`] = `
"[
  {
    \\"name\\": \\"@company/package-with-externals\\",
    \\"external\\": false,
    \\"files\\": [
      {
        \\"name\\": \\"./packages/with-externals-imports/src/other.js\\",
        \\"hasTypes\\": false,
        \\"edges\\": [
          {
            \\"packageName\\": \\"@ember/object\\",
            \\"hasTypes\\": false,
            \\"fileName\\": \\"@ember/object\\",
            \\"missing\\": true
          },
          {
            \\"packageName\\": \\"my-external-with-types\\",
            \\"hasTypes\\": true,
            \\"fileName\\": \\"./packages/with-externals-imports/node_modules/my-external-with-types/addon/index.ts\\",
            \\"missing\\": false
          }
        ]
      },
      {
        \\"name\\": \\"./packages/with-externals-imports/src/index.mts\\",
        \\"hasTypes\\": true,
        \\"edges\\": [
          {
            \\"packageName\\": \\"@company/package-with-externals\\",
            \\"hasTypes\\": false,
            \\"fileName\\": \\"./packages/with-externals-imports/src/other.js\\",
            \\"missing\\": false
          },
          {
            \\"packageName\\": \\"my-external-addon-no-types\\",
            \\"hasTypes\\": false,
            \\"fileName\\": \\"./packages/with-externals-imports/node_modules/my-external-addon-no-types/addon/index.js\\",
            \\"missing\\": false
          },
          {
            \\"packageName\\": \\"my-external\\",
            \\"hasTypes\\": true,
            \\"fileName\\": \\"./packages/with-externals-imports/node_modules/my-external/dist/index.js\\",
            \\"missing\\": false
          },
          {
            \\"packageName\\": \\"@company/package-e\\",
            \\"hasTypes\\": true,
            \\"fileName\\": \\"./packages/package-e/src/foo/index.ts\\",
            \\"missing\\": true
          }
        ]
      }
    ]
  },
  {
    \\"name\\": \\"@company/package-e\\",
    \\"external\\": false,
    \\"files\\": [
      {
        \\"name\\": \\"./packages/package-e/src/foo/index.ts\\",
        \\"hasTypes\\": true,
        \\"edges\\": []
      }
    ]
  },
  {
    \\"name\\": \\"my-external\\",
    \\"external\\": true,
    \\"files\\": [
      {
        \\"name\\": \\"./packages/with-externals-imports/node_modules/my-external/dist/index.js\\",
        \\"hasTypes\\": true,
        \\"edges\\": []
      }
    ]
  },
  {
    \\"name\\": \\"my-external-addon-no-types\\",
    \\"external\\": true,
    \\"files\\": [
      {
        \\"name\\": \\"./packages/with-externals-imports/node_modules/my-external-addon-no-types/addon/index.js\\",
        \\"hasTypes\\": false,
        \\"edges\\": []
      }
    ]
  },
  {
    \\"name\\": \\"my-external-with-types\\",
    \\"external\\": true,
    \\"files\\": [
      {
        \\"name\\": \\"./packages/with-externals-imports/node_modules/my-external-with-types/addon/index.ts\\",
        \\"hasTypes\\": true,
        \\"edges\\": []
      }
    ]
  },
  {
    \\"name\\": \\"@ember/object\\",
    \\"external\\": true,
    \\"files\\": [
      {
        \\"name\\": \\"@ember/object\\",
        \\"hasTypes\\": false,
        \\"edges\\": []
      }
    ]
  }
]"
`;
